libtenzir/builtins/functions/time.cpp

```diff
@@ -240,55 +241,111 @@ class year_month_day final : public function_plugin {
                 if (arg.IsNull(i)) {
                   check(b.AppendNull());
                   continue;
                 }
-                auto val = arg.Value(i);
-                auto pre = static_cast<double>(val / factor);
-                auto post = static_cast<double>(val % factor) / factor;
-                check(b.Append(pre + post));
+                auto&& value = value_at(time_type{}, arg, i);
+                auto duration_since_day_start
+                  = value - std::chrono::floor<std::chrono::days>(value);
+                auto hours = std::chrono::duration_cast<std::chrono::hours>(
+                  duration_since_day_start);
+                duration_since_day_start -= hours;
+                auto minutes = std::chrono::duration_cast<std::chrono::minutes>(
+                  duration_since_day_start);
+                duration_since_day_start -= minutes;
+                // Convert remaining duration to seconds with fractional part
+                auto seconds_double
+                  = static_cast<double>(duration_since_day_start.count()) / 1e9;
+                check(b.Append(seconds_double));
               }
               return series{double_type{}, finish(b)};
-            },
-            [&](const auto&) {
-              diagnostic::warning("`{}` expected `duration`, but got `{}`",
-                                  name(), arg.type.kind())
-                .primary(expr)
-                .emit(ctx);
-              return series::null(double_type{}, arg.length());
-            },
-          };
-          return match(*arg.array, f);
-        });
+            }
+            // For hour and minute, use Int64Builder
+            auto b = arrow::Int64Builder{};
+            check(b.Reserve(arg.length()));
+            for (auto i = int64_t{0}; i < arg.length(); ++i) {
+              if (arg.IsNull(i)) {
+                check(b.AppendNull());
+                continue;
+              }
+              auto&& value = value_at(time_type{}, arg, i);
+              auto duration_since_day_start
+                = value - std::chrono::floor<std::chrono::days>(value);
+              auto result = int64_t{0};
+              switch (hms_subtype_) {
+                case hms_subtype::hour: {
+                  auto hours = std::chrono::duration_cast<std::chrono::hours>(
+                    duration_since_day_start);
+                  result = hours.count();
+                  break;
+                }
+                case hms_subtype::minute: {
+                  auto hours = std::chrono::duration_cast<std::chrono::hours>(
+                    duration_since_day_start);
+                  duration_since_day_start -= hours;
+                  auto minutes
+                    = std::chrono::duration_cast<std::chrono::minutes>(
+                      duration_since_day_start);
+                  result = minutes.count();
+                  break;
+                }
+                default:
+                  break;
```

Comment:
```suggestion
                  TENZIR_UNREACHABLE();
```

Even better would be an non-defaulted switch in lambda of course, but whatever.

Reply:
I'll change this, actually. One more thing to point out in my `.rules`: avoid default statements when using switch.